src/components/appSwitch/AppSwitch.js

```diff
@@ -258,8 +217,37 @@ const AppSwitch = (props: LoadingProps) => {
 
   useEffect(() => {
     // initialize with initRegistered = true only if user is loggedin correctly (ie jwt not expired)
-    initWalletAndStorage(undefined, 'SEED', isLoggedIn).then(() => log.debug('storage and wallet ready'))
-  }, [initWalletAndStorage, isLoggedIn])
+    const run = async () => {
+      if (initializing.current) {
+        return
+      }
+      try {
+        initializing.current = true
+        await initWalletAndStorage(undefined, 'SEED')
+        log.debug('storage and wallet ready')
+      } catch (e) {
+        if ('UnsignedJWTError' === e.name) {
+          return restartWithMessage(
+            "You haven't used GoodDollar app on this device for a long time. " +
+              'You need to sign in again. Make sure to use the same account you previously signed in with.',
+          )
+        }
+
+        // if error in realmdb logout the user, he needs to signin/signup again
+        log.error('failed initializing app', e.message, e)
+
+        if (e.message.includes('realmdb')) {
+          return restartWithMessage(
+            'We are sorry, but due to database upgrade, you need to perform the Signup process again. ' +
+              'Make sure to use the same account you previously signed in with.',
+          )
+        }
+
+        restartWithMessage('Wallet could not be loaded. Please refresh.', false)
+      }
+    }
+    run()
```

Comment:
1. In this effect we're calling just one async fn, so it could be written as promise chain
2. only .catch block could be moved to the separate fn
3. need to reset initializing flag in finally 

```suggestion
    if (initializing.current) {
      return
    }
    
    const onInitializationFailed = e => {
      if ('UnsignedJWTError' === e.name) {
        return restartWithMessage(
          "You haven't used GoodDollar app on this device for a long time. " +
            'You need to sign in again. Make sure to use the same account you previously signed in with.',
        )
      }

      // if error in realmdb logout the user, he needs to signin/signup again
      log.error('failed initializing app', e.message, e)

      if (e.message.includes('realmdb')) {
        return restartWithMessage(
          'We are sorry, but due to database upgrade, you need to perform the Signup process again. ' +
            'Make sure to use the same account you previously signed in with.',
        )
      }

      restartWithMessage('Wallet could not be loaded. Please refresh.', false)
    }
    
    initializing.current = true
    initWalletAndStorage(undefined, 'SEED')
      .then(() => log.debug('storage and wallet ready'))
      .catch(onInitializationFailed)
      .finally(() => (initializing.current = false))    
```